components/raftstore/src/store/fsm/store.rs

```diff
@@ -2656,39 +2662,52 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> StoreFsmDelegate<'a, EK, ER
         self.ctx.router.report_status_update()
     }
 
-    fn on_create_peer(&self, region: Region) {
-        info!("creating a peer"; "peer" => ?region);
-        let mut kv_wb = self.ctx.engines.kv.write_batch();
-        let region_state_key = keys::region_state_key(region.get_id());
-        match self
-            .ctx
-            .engines
-            .kv
-            .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
+    fn on_unsafe_recovery_create_peer(&self, region: Region) {
+        info!("Unsafe recovery, creating a peer"; "peer" => ?region);
         {
-            Ok(Some(region_state)) => {
-                info!(
-                    "target region already exists, existing region: {:?}, want to create: {:?}",
-                    region_state, region
-                );
-                return;
-            }
-            Ok(None) => {}
-            Err(e) => {
-                panic!("cannot determine whether {:?} exists, err {:?}", region, e)
+            let meta = self.ctx.store_meta.lock().unwrap();
+            if let Some((_, id)) = meta
+                .region_ranges
+                .range((
+                    Excluded(data_key(region.get_start_key())),
+                    Unbounded::<Vec<u8>>,
+                ))
+                .next()
+            {
+                let exist_region = &meta.regions[id];
+                if enc_start_key(exist_region) < data_end_key(region.get_end_key()) {
+                    if exist_region.get_id() == region.get_id() {
+                        warn!("Unsafe recovery, region has already been created."; "region"=>?region);
+                        return;
+                    } else {
+                        error!(
+                            "Unsafe recovery, region to be created overlaps with an existing region";
+                            "region" => ?region,
+                            "exist_region" => ?exist_region,
+                        );
+                        return;
+                    }
+                }
             }
-        };
-        peer_storage::write_peer_state(&mut kv_wb, &region, PeerState::Normal, None)
-            .unwrap_or_else(|e| {
-                panic!(
-                    "fail to add peer state into write batch while creating {:?} err {:?}",
-                    region, e
-                )
-            });
+        }
+        let mut kv_wb = self.ctx.engines.kv.write_batch();
```

Comment:
call delete_all_in_range to clean the stale data

Reply:
Added, but should we clean stale data here? I expected we clean it through delete.

Reply:
Yes, just in case